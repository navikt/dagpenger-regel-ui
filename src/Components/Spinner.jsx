import React from 'react';
import NavFrontendSpinner from 'nav-frontend-spinner';

const Spinner = props => (
  <div className="flex">
    <NavFrontendSpinner {...props} />
  </div>
);

export default Spinner;
