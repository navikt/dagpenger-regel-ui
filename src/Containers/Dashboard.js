import React, { useState, useEffect } from 'react';
import axios from 'axios';

const Dashboard = (
  location,
) => {
  const routeParams = location.match.params;
  const b = 'dsfdsfdsfd';
  const { aktorId, beregningdato, vilkarid } = routeParams;

  const [data, setData] = useState({ inntekter: {} });

  useEffect(async () => {
    const result = await axios(
      'http://localhost:3000/mock/inntekter.json',
    );

    setData(result.data);
  });

  return (
    <div />
  );
};

export default Dashboard;
