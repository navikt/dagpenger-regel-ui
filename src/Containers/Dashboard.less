@import '~nav-frontend-core/less/_variabler';

@width: 300px;
@height: 300px;

.grid {
  display: flex;
  flex-direction: row;
  justify-content: start;
  overflow-x: auto;
  width: 100%;

  .item {
    padding: 16px;
    border-right: 1px solid black;
    border-top: 1px solid black;
  }

  .arbeidsgivere {
    background: @navLysGra;
    width: @width;
    position: absolute;
    left: 0;

    .arbeidsgiver {
      height: @height;
    }
  }

  .maaneder {
    display: flex;
    flex-direction: row;
    justify-content: start;
    overflow-x: auto;
    margin-left: @width;
    width: 100%;

    .maaned {
      background: white;
      width: @width;

      .maanednavn {
        background: @navMorkGra;
        color: white;
      }

      .inntekter {
        height: @height;

        .inntekt {

        }
      }
    }
  }
}
