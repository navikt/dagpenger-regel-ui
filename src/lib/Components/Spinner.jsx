import React from "react";
import NavFrontendSpinner from "nav-frontend-spinner";

function Spinner(props) {
  return (
    <div className="flex center">
      <NavFrontendSpinner {...props} />
    </div>
  );
}

export default Spinner;
