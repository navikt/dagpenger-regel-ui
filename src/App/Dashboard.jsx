import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Knapp } from 'nav-frontend-knapper';
import { useQuery } from '@apollo/react-hooks';
import Panel from 'nav-frontend-paneler';
import { nb } from 'date-fns/locale';
import AlertStripe from 'nav-frontend-alertstriper';
import { formatDistance } from 'date-fns';
import { Normaltekst, Ingress, Element } from 'nav-frontend-typografi';
import { loader } from 'graphql.macro';
import { OkAvbrytModal } from '../Components/OkAvbrytModal';
import { ReactComponent as MannIkon } from '../images/mann.svg';
import { ReactComponent as KvinneIkon } from '../images/kvinne.svg';
import InntektsForm from '../Inntekt/InntektsForm';
import EditedIkon from '../Components/EditedIkon';
import Spinner from '../Components/Spinner';
import Spacer from '../Components/Spacer';
import { DDMMYYYYHHMM_FORMAT } from '../Utils/datoFormat';
import { formatDato, eachMonthOfInterval } from '../Utils/datoUtils';

const GET_INNTEKT = loader('./GET_INNTEKT.gql');

const getKjønn = kjoenn => {
  if (kjoenn === 'Kvinne') {
    return <KvinneIkon />;
  }

  return <MannIkon />;
};

const getAlleMåneder = (fraDato, tilDato) => {
  const måneder = eachMonthOfInterval({
    start: new Date(fraDato),
    end: new Date(tilDato),
  });

  return måneder;
};

const set36Måneder = data => {
  const { fraDato, tilDato } = {
    fraDato: '2016-07',
    tilDato: '2019-06',
  };
  if (fraDato && tilDato) {
    const måneder = getAlleMåneder(fraDato, tilDato);

    data.map(a => {
      if (a.posteringer === undefined) {
        Object.assign(a, {
          posteringer: {},
        });
      }

      måneder.forEach(måned => {
        if (!a.posteringer[måned]) {
          // console.log('if');
          Object.assign(a.posteringer, {
            [måned]: [],
          });
        }
      });
      return data;
    });
  }
  return data;
};

// TODO generaliser slik at den kan fortsatt være i utils
const groupBy = (list, keyGetter) => {
  const map = new Map();
  list.forEach(item => {
    const key = keyGetter(item);
    const collection = map.get(key);
    const dato = `${item.periode.year}-${item.periode.month}`;
    if (!collection) {
      map.set(key, {
        ...item.virksomhet,
        posteringer: { [dato]: [item] },
      });
    } else if (collection.posteringer[dato]) {
      collection.posteringer[dato].push(item);
    } else {
      Object.assign(collection.posteringer, {
        [dato]: [item],
      });
    }
  });
  return Array.from(map.values());
};

const Dashboard = ({ readOnly, location }) => {
  const inntektRequest = new URLSearchParams(location.search);

  const { data, error, loading } = useQuery(GET_INNTEKT, {
    variables: {
      personId: inntektRequest.get('aktorId'),
      vedtakId: inntektRequest.get('vedtakId'),
      beregningsdato: inntektRequest.get('beregningdato'),
    },
  });

  const [hentInntektStatus, setHentInntekttatus] = useState(false);
  const [isHentInntektModalOpen, setHentInntektModal] = useState(false);

  const fetchUncachedInntekt = async () => {
    setHentInntekttatus('fetching');

    setHentInntekttatus(true);
  };

  if (loading) {
    return <Spinner type="XL" />;
  }

  if (error || !data.person) {
    return <>error</>;
  }

  const måneder = getAlleMåneder('2016-07', '2019-06');

  const { person } = data;

  const inntekter =
    !loading &&
    set36Måneder(groupBy(person.vedtak.inntekt.posteringer, postering => postering.virksomhet.organisasjonsnummer || postering.virksomhet.naturligIdent));
  console.log('data', data, inntekter);
  return (
    <>
      <Panel border>
        <div className="flex">
          {person.naturligIdent && (
            <>
              <div className="marginhoyre16">{getKjønn(person.kjoenn)}</div>
              <div>
                <Ingress>{person.navn}</Ingress>
                <Normaltekst>Fødselsnummer</Normaltekst>
                <Ingress>{person.naturligIdent}</Ingress>
              </div>
            </>
          )}
          <div className="flexend flex noprint">
            {person.manueltRedigert && (
              <div className="marginhoyre16 flex">
                <EditedIkon />
                <Element>Manuelt redigert</Element>
              </div>
            )}
          </div>
        </div>
      </Panel>

      <Spacer sixteenPx />

      {hentInntektStatus && (
        <div aria-live="polite">
          <AlertStripe type="info">Inntekt innhentet. Trykk bekreft for å lagre.</AlertStripe>
          <Spacer sixteenPx />
        </div>
      )}

      <div className="flex hentinntekter">
        <Knapp onClick={() => setHentInntektModal(true)} autoDisableVedSpinner disabled={readOnly} spinner={hentInntektStatus === 'fetching'}>
          Hent inntekter på nytt
        </Knapp>
        <div className="marginvenstre16">
          Opplysninger hentet :
          <Normaltekst>
            {formatDato(new Date(person.vedtak.inntekt.timestamp), DDMMYYYYHHMM_FORMAT)}
            {', '}
            <b>{formatDistance(new Date(person.vedtak.inntekt.timestamp), new Date(), { locale: nb, addSuffix: true })}</b>
          </Normaltekst>
        </div>
        <OkAvbrytModal
          isOpen={isHentInntektModalOpen}
          text="Når du henter inn nyeste inntekt fra skatt så vil alle tidligere endringene gå tapt."
          avbrytCallback={() => setHentInntektModal(false)}
          OkCallback={() => {
            fetchUncachedInntekt();
            setHentInntektModal(false);
          }}
        />
      </div>

      <InntektsForm
        initialValues={{
          person: {
            ...person,
            vedtak: {
              ...person.vedtak,
              inntekt: {
                ...person.vedtak.inntekt,
                posteringer: [...inntekter],
              },
            },
          },
        }}
        hentInntektStatus={hentInntektStatus}
        måneder={måneder}
        readOnly={readOnly}
      />
    </>
  );
};

Dashboard.propTypes = {
  readOnly: PropTypes.bool,
  location: PropTypes.shape().isRequired,
};

Dashboard.defaultProps = {
  readOnly: false,
};

export default Dashboard;
