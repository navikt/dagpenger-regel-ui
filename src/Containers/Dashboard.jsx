import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { Formik } from 'formik';
import axios from 'axios';
import { Knapp } from 'nav-frontend-knapper';
import AlertStripe from 'nav-frontend-alertstriper';
import { Normaltekst, Ingress } from 'nav-frontend-typografi';
import Panel from 'nav-frontend-paneler';
import { ToggleKnapp } from 'nav-frontend-toggle';
import Spinner from '../Components/Spinner';
import Spacer from '../Components/Spacer';
import { Inntektstabell } from './Inntektstabell';
import { OkAvbrytModal } from '../Components/OkAvbrytModal';
import {
  getInntekt, getUncachedInntekt, lagreInntekt, getName,
} from '../lib/inntektApiClient';
import { ReactComponent as MannIkon } from '../images/mann.svg';

import './Dashboard.css';

export const findArbeidsgivere = (inntekt) => {
  const map = new Map();
  inntekt.arbeidsInntektMaaned
    .forEach(mnd => mnd.arbeidsInntektInformasjon.inntektListe
      .forEach((arbeidsgiver) => {
        // TODO fikse denne, bør flyttes til backend ved et senere tidspunkt
        const navn = 'test'; // getName(JSON.stringify({ id: arbeidsgiver.virksomhet.identifikator, aktørType: arbeidsgiver.virksomhet.aktoerType }));

        map.set(arbeidsgiver.virksomhet.identifikator, { navn, ...arbeidsgiver.virksomhet });
      }));

  return Array.from(map.values())
    .sort((a, b) => b.identifikator - a.identifikator);
};


const inntektRequest = queryParams => ({
  aktørId: queryParams.get('aktorId'),
  vedtakId: queryParams.get('vedtakId'),
  beregningsDato: queryParams.get('beregningdato'),
});

const Dashboard = ({ readOnly, location }) => {
  const [inntektdata, setInntektdata] = useState({ inntektId: '', inntekt: { arbeidsInntektMaaned: [], ident: {}, manueltRedigert: false } });
  const [arbeidsgivere, setArbeidsgivere] = useState([]);
  const [hentInntektStatus, setHentInntekttatus] = useState(false);
  const [isHentInntektModalOpen, setHentInntektModal] = useState(false);


  useEffect(() => {
    const getInntektFromApi = async () => {
      let result;
      if (process.env.NODE_ENV !== 'production') {
        result = await axios(
          `${process.env.PUBLIC_URL}/mock/mock.json`,
        );
      } else {
        result = await getInntekt(inntektRequest(new URLSearchParams(location.search)));
      }
      setInntektdata({ ...result.data });
      setArbeidsgivere(findArbeidsgivere(result.data.inntekt));
    };

    getInntektFromApi();
  }, [location.search]);

  const fetchUncachedInntekt = async () => {
    setHentInntekttatus('fetching');
    if (process.env.NODE_ENV !== 'production') {
      const result = await axios(
        `${process.env.PUBLIC_URL}/mock/mock1.json`,
      );
      setInntektdata({ ...result.data });
      setArbeidsgivere(findArbeidsgivere(result.data.inntekt));
      setHentInntekttatus(true);
    } else {
      getUncachedInntekt(inntektRequest(new URLSearchParams(location.search)))
        .then((result) => {
          setInntektdata({ ...result.data });
          setArbeidsgivere(findArbeidsgivere(result.data.inntekt));
          setHentInntekttatus(true);
        })
        // eslint-disable-next-line no-unused-vars
        .catch((error) => {
          setHentInntekttatus('error');
        });
    }
  };

  if (!arbeidsgivere.length || !inntektdata.inntekt.arbeidsInntektMaaned.length) {
    return <Spinner type="XL" />;
  }

  return (
    <>
      <Panel border>
        <div className="flex">
          <MannIkon />
          <div className="paddingleft">
            <Ingress>Ola Nordmann</Ingress>
            <Normaltekst>{`AktørId: ${inntektdata.inntekt.ident.identifikator}`}</Normaltekst>
          </div>
          <div className="flexend hoyre">
            InntektId
            <Normaltekst>{inntektdata.inntektId.id}</Normaltekst>
          </div>
        </div>
      </Panel>

      <Spacer sixteenPx />

      {hentInntektStatus
      && <div aria-live="polite"><AlertStripe type="info">Ny inntekt hentet</AlertStripe></div>}
      {hentInntektStatus === 'error'
      && (
        <div aria-live="polite">
          <AlertStripe type="feil">En feil skjedde under henting av frisk inntekt</AlertStripe>
        </div>
      )}

      <div className="flex">
        <Knapp
          onClick={() => setHentInntektModal(true)}
          autoDisableVedSpinner
          spinner={hentInntektStatus === 'fetching'}
        >

        Hent inntekter på nytt
        </Knapp>

        <OkAvbrytModal
          isOpen={isHentInntektModalOpen}
          text="Når du henter inn nyeste inntekt fra skatt så vil alle endringene dine gå tapt."
          avbrytCallback={() => setHentInntektModal(false)}
          OkCallback={() => {
            fetchUncachedInntekt();
            setHentInntektModal(false);
          }}
        />
        <div className="flexend">
          <ToggleKnapp>Vis alle</ToggleKnapp>
        </div>

      </div>


      <Spacer sixteenPx />

      <Formik
        validate
        enableReinitialize
        initialValues={{
          ...inntektdata,
          arbeidsgivere: [
            ...arbeidsgivere,
          ],
        }}
        onSubmit={(values, actions) => {
          setTimeout(() => {
            lagreInntekt(values)
              .then((result) => {
                setInntektdata({ ...result.data });
                setArbeidsgivere(findArbeidsgivere(result.data.inntekt));
                actions.setStatus({ success: true });
              })
              // eslint-disable-next-line no-unused-vars
              .catch((error) => {
                actions.setStatus({ failure: true });
              }).finally(() => {
                actions.setSubmitting(false);
              });
          }, 1000);
        }}
        render={formProps => (
          <Inntektstabell
            {...formProps}
            readOnly={readOnly}
            hentInntektStatus={hentInntektStatus}
          />
        )}
      />

    </>
  );
};

Dashboard.propTypes = {
  location: PropTypes.shape({ search: PropTypes.string.isRequired }).isRequired,
  readOnly: PropTypes.bool,
};

Dashboard.defaultProps = { readOnly: false };

export default Dashboard;
