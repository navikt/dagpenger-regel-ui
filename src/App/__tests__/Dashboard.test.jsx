import React from 'react';
import renderer from 'react-test-renderer';
import { MockedProvider } from '@apollo/react-testing';
import { loader } from 'graphql.macro';
import Spinner from '../../Components/Spinner';
import Dashboard from '../Dashboard';

const GET_INNTEKT = loader('../../Graphql/GET_INNTEKT.gql');

const location = {
  search: 'http://localhost/inntekter/?aktorId=3_G_INNTEKT&vedtakId=12345&beregningdato=2019-07-01',
};

const inntektRequest = queryParams => ({
  personId: queryParams.get('aktorId'),
  vedtakId: queryParams.get('vedtakId'),
  beregningsdato: queryParams.get('beregningdato'),
});

const mocks = [
  {
    request: {
      query: GET_INNTEKT,
      variables: { ...inntektRequest(new URLSearchParams(location.search)) },
    },
    result: {
      data: {
        dog: { id: '1', name: 'Buck', breed: 'bulldog' },
      },
    },
  },
];

describe('Dashboard', () => {
  test('Skal vise ny Dashboard med loader', () => {
    const component = renderer.create(
      <MockedProvider mocks={[]} addTypename={false}>
        <Dashboard location={location} />
      </MockedProvider>,
    );
    const tree = component.toJSON();

    const spinner = renderer.create(<Spinner type="XL" />);

    expect(tree).toEqual(spinner.toJSON());
  });

  test('Skal vise ny Dashboard', () => {
    const component = renderer.create(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Dashboard location={location} />
      </MockedProvider>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
