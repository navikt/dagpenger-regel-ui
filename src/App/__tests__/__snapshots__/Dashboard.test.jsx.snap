// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Skal vise ny Dashboard 1`] = `
<div
  className="flex"
>
  <svg
    aria-label="Laster innhold"
    className="spinner spinner--xl"
    focusable="false"
    height={32}
    kind="spinner"
    preserveAspectRatio="xMidYMid"
    viewBox="0 0 50 50"
    width={32}
  >
    <title>
      Venter...
    </title>
    <circle
      cx="25"
      cy="25"
      fill="none"
      r="20"
      stroke="#bdbab7"
      strokeWidth="5"
      xmlns="http://www.w3.org/2000/svg"
    />
    <circle
      cx="25"
      cy="25"
      fill="none"
      r="20"
      stroke="#7f756c"
      strokeDasharray="50 155"
      strokeLinecap="round"
      strokeWidth="5"
    />
  </svg>
</div>
`;

exports[`Dashboard Skal vise ny Dashboard med loader 1`] = `
<div
  className="flex center"
>
  <svg
    aria-label="Laster innhold"
    className="spinner spinner--xl"
    focusable="false"
    height={32}
    kind="spinner"
    preserveAspectRatio="xMidYMid"
    viewBox="0 0 50 50"
    width={32}
  >
    <title>
      Venter...
    </title>
    <circle
      cx="25"
      cy="25"
      fill="none"
      r="20"
      stroke="#bdbab7"
      strokeWidth="5"
      xmlns="http://www.w3.org/2000/svg"
    />
    <circle
      cx="25"
      cy="25"
      fill="none"
      r="20"
      stroke="#7f756c"
      strokeDasharray="50 155"
      strokeLinecap="round"
      strokeWidth="5"
    />
  </svg>
</div>
`;
